morethuente.go

```diff
@@ -0,0 +1,384 @@
+		if mt.step <= mt.lower || mt.step >= mt.upper || mt.upper-mt.lower <= mt.StepTolerance*mt.upper {
+			// If further progress is not possible, set step to the best step
+			// obtained during the search.
+			mt.step = mt.x
+		}
+	}
+
+	return FuncEvaluation | GradEvaluation, mt.step, nil
+}
+
+// nextStep computes the next safeguarded step and updates the interval that
+// contains a step that satisfies the sufficient decrease and curvature
+// conditions.
+func (mt *MoreThuente) nextStep(fx, gx, fy, gy, f, g float64) {
+	x := mt.x
+	y := mt.y
+	step := mt.step
+
+	gSign := g
+	if gx < 0 {
+		gSign = -g
+	}
+
+	var next float64
+	var bracketed bool
+	switch {
+	case f > fx:
+		// A higher function value. The minimum is bracketed between x and step.
+		// We want the next step to be closer to x because the function value
+		// there is lower.
+
+		theta := 3*(fx-f)/(step-x) + gx + g
+		s := math.Max(math.Abs(gx), math.Abs(g))
+		s = math.Max(s, math.Abs(theta))
+		gamma := s * math.Sqrt((theta/s)*(theta/s)-(gx/s)*(g/s))
+		if step < x {
+			gamma = -gamma
+		}
+		p := (gamma - gx) + theta
+		q := ((gamma - gx) + gamma) + g
+		r := p / q
+		stpc := x + r*(step-x)
+		stpq := x + ((gx/((fx-f)/(step-x)+gx))/2)*(step-x)
+
+		if math.Abs(stpc-x) < math.Abs(stpq-x) {
+			// The cubic step is closer to x than the quadratic step.
+			// Take the cubic step.
+			next = stpc
+		} else {
+			// If f is much larger than fx, then the quadratic step may be too
+			// close to x. Therefore heuristically take the average of the
+			// cubic and quadratic steps.
+			next = stpc + (stpq-stpc)/2
+		}
+		bracketed = true
+
+	case gSign < 0:
+		// A lower function value and derivatives of opposite sign. The minimum
+		// is bracketed between x and step. If we choose a step that is far
+		// from step, the next iteration will also likely fall in this case.
+
+		theta := 3*(fx-f)/(step-x) + gx + g
+		s := math.Max(math.Abs(gx), math.Abs(g))
+		s = math.Max(s, math.Abs(theta))
+		gamma := s * math.Sqrt((theta/s)*(theta/s)-(gx/s)*(g/s))
+		if step > x {
+			gamma = -gamma
+		}
+		p := (gamma - g) + theta
+		q := ((gamma - g) + gamma) + gx
+		r := p / q
+		stpc := step + r*(x-step)
+		stpq := step + (g/(g-gx))*(x-step)
+
+		if math.Abs(stpc-step) > math.Abs(stpq-step) {
+			// The cubic step is farther from x than the quadratic step.
+			// Take the cubic step.
+			next = stpc
+		} else {
+			// Take the quadratic step.
+			next = stpq
+		}
+		bracketed = true
+
+	case math.Abs(g) < math.Abs(gx):
+		// A lower function value, derivatives of the same sign, and the
+		// magnitude of the derivative decreases. Extrapolate function values
+		// at x and step so that the next step lies between step and y.
+
+		theta := 3*(fx-f)/(step-x) + gx + g
+		s := math.Max(math.Abs(gx), math.Abs(g))
+		s = math.Max(s, math.Abs(theta))
+		gamma := s * math.Sqrt(math.Max(0, (theta/s)*(theta/s)-(gx/s)*(g/s)))
+		if step > x {
+			gamma = -gamma
+		}
+		p := (gamma - g) + theta
```

Comment:
They do literally nothing though here, I’m pretty sure. The Go spec guarantees they happen in the order as written. What is says:

"Floating-point operations within a single expression are evaluated according to the associativity of the operators.  Explicit parentheses affect the evaluation by overriding the default associativity. In the expression x + (y + z) the addition y + z is performed before adding x. “

So, by default, gamma - g + theta  IS (gamma - g) + theta. I’m not sure what the Fortran rules of this are.


> On Jan 21, 2016, at 7:04 PM, Vladimír Chalupecký <notifications@github.com> wrote:
> 
> In morethuente.go <https://github.com/gonum/optimize/pull/148#discussion_r50494427>:
> 
> > +		}
> > +		bracketed = true
> > +
> > +	case math.Abs(g) < math.Abs(gx):
> > +		// A lower function value, derivatives of the same sign, and the
> > +		// magnitude of the derivative decreases. Extrapolate function values
> > +		// at x and step so that the next step lies between step and y.
> > +
> > +		theta := 3*(fx-f)/(step-x) + gx + g
> > +		s := math.Max(math.Abs(gx), math.Abs(g))
> > +		s = math.Max(s, math.Abs(theta))
> > +		gamma := s * math.Sqrt(math.Max(0, (theta/s)*(theta/s)-(gx/s)*(g/s)))
> > +		if step > x {
> > +			gamma = -gamma
> > +		}
> > +		p := (gamma - g) + theta
> Yes, that's what I thought too. The Fortran original uses these "unnecessary" parentheses in the interpolation code consistently enough for me to think that there is a purpose (i.e., floating point magic) behind it so I kept them. I will try if removing them has any effect on the tests. At any case, I'll be happy to remove them if you want.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/gonum/optimize/pull/148/files#r50494427>.
>